File: r/src/player/song.r
source("src/player/utility.r")

PlayerSong <- setClass(
    "PlayerSong",
    slots = c(
        title="character",
        artist="character",
        rating="numeric"
    )
)

# Definitions
setGeneric(name="printSong", def=function(self) { standardGeneric("printSong")})

# Implementations
setMethod(f="printSong", signature="PlayerSong", definition=function(self) {
    title <- playerPrintWithBuffer(self@title, 20)
    artist <- playerPrintWithBuffer(self@artist, 27)
    message(title, " ", artist, " ", self@rating)
})

File: rust/src/player/song.rs
use crate::player::utility;

/* 
  Struct fields have to be public for us to do
  Song{artist:..., } from another file
  */
pub struct Song {
    pub title: String,
    pub artist: String,
    pub rating: u8
}

impl Song {
    pub fn print(&self) {
        let title = utility::print_with_buffer(self.title.as_str(), 20);
        let artist = utility::print_with_buffer(self.artist.as_str(), 27);
        println!("{} {} {}", title, artist, self.rating);
    }
}